tensorzero-core/src/observability.rs

```diff
@@ -37,37 +130,160 @@ fn internal_build_otel_layer<T: SpanExporter + 'static>(
     );
 
     if let Some(exporter) = override_exporter {
-        provider = provider.with_simple_exporter(exporter);
+        builder = builder.with_simple_exporter(exporter);
     } else {
-        provider = provider.with_batch_exporter(
-            opentelemetry_otlp::SpanExporter::builder()
-                .with_tonic()
-                .build()
-                .map_err(|e| {
-                    Error::new(ErrorDetails::Observability {
-                        message: format!("Failed to create OTLP exporter: {e}"),
-                    })
-                })?,
-        );
+        builder = builder.with_batch_exporter(exporter);
     }
-    let provider = provider.build();
+    let provider = builder.build();
+
     let tracer = provider.tracer("tensorzero");
+    Ok((provider, tracer))
+}
+
+impl Tracer for TracerWrapper {
+    type Span = <SdkTracer as Tracer>::Span;
+
+    // This is the only method where we dispatch to a different `Tracer` - all other methods
+    // just forward to `default_tracer`/`default_provider`.
+    // This is fine, since `build_with_context` is the only method used by `tracing-opentelemetry`
+    // when building an OpenTelemetry span from a `tracing::Span`.
+    fn build_with_context(
+        &self,
+        builder: opentelemetry::trace::SpanBuilder,
+        parent_cx: &opentelemetry::Context,
+    ) -> Self::Span {
+        if let Some(key) = parent_cx.get::<CustomTracerKey>() {
+            let tracer = self.custom_tracers.try_get_with_by_ref(key, || {
+                // We need to provide a dummy generic parameter to satisfy the compiler
```

Comment:
please add a comment about cache churn here and that we may want to benchmark this at some point!